Adding container build option, documentation

elai-shalev opened the pull request:
Main additions:

1. Dockerfile to build container image of project 
2. build script to build and push the image 
3. supervisor config file
4. nginx config
5. deployment readme 
6. changes in consuming env files to django settings

elai-shalev commented on api/views.py:
```diff
@@ -38,7 +38,18 @@ def analyze_address(request):
-        street_data = json.load(f)
-    street_names = street_data.get("t_rechov_values", [])
-    return JsonResponse({"streets": street_names})
+    try:
```
That file is is hardcoded in the project

elai-shalev commented on tamaod/settings.py:
```diff
@@ -11,16 +11,17 @@
 
-ENVIRONMENT = os.getenv('DJANGO_ENV', 'production')
-
+# Loading environment type, development by default
```
@tinez wdyt?

tinez replied:
Generally, we should not have `if ENVIRONMENT == production` in the code. Instead, we should provide reasonable defaults for all envs and allow the user to override.
Re: `SECRET_KEY` - we already provide an example how to generate one in `SECURITY.md`, so let's use that snippet if none was provided by the user.
`DEBUG` should be always set to false by default.
I like the `get_bool` helper that was suggested.

elai-shalev replied:
Done

tinez commented on deploy/supervisord.conf:
```diff
@@ -0,0 +1,25 @@
+autorestart=true
+stdout_logfile=/var/log/supervisor/nginx_stdout.log
+stderr_logfile=/var/log/supervisor/nginx_stderr.log
+user=root
```
We should not be running as `root` but as the `app` user that we create in the dockerfile

tinez commented on supervisord.conf:
```diff
@@ -0,0 +1,25 @@
+autorestart=true
+stdout_logfile=/var/log/supervisor/django_stdout.log
+stderr_logfile=/var/log/supervisor/django_stderr.log
+user=root
```
Same here

tinez commented on tamaod/urls.py:
```diff
@@ -16,10 +16,16 @@
+    path('', include('ui.urls')),
 ]
+
+if settings.DEBUG:
```
Hmm I'm not sure what these piece of code does and why we're using it only in `DEBUG`

elai-shalev commented:
@tinez 
Big commits:
1. Fixed .env file loading -- users need to pass the env file to run locally or in container, defaults to .env.prod if avaliable
2. pytests default to .env.test if avaliable 
3. Added ssl, and ssl bypass option for running in dev mode on https enforcing browser
4. created the container file, deployment.md guide, built and pushed the image, added script to do so also 
5. basic app function should work in dev mode and in container

Let me know what you think

tinez commented on api/services/real.py:
```diff
@@ -5,64 +5,98 @@
+
+        # Nominatim requires a User-Agent header (usage policy)
+        # Set defaults if not provided
+        user_agent = os.getenv('USER_AGENT', 'TamaOd/1.0 (https://github.com/your-org/tamaod)')
```
Here IMO it would make sense to pass the real github url to the project, so that the Nominatim guys can figure out who's responsible ;)

tinez commented on api/services/real.py:
```diff
@@ -5,64 +5,98 @@
+
+        # Build headers dict, only including non-None values
+        headers = {}
+        if user_agent:
```
We have defaults for `user_agent` and `referer`, so these `if`s are not really necessary. We can still end up with the user setting e.g. `USER_AGENT` to an empty string, but this should be handled above:
```python
user_agent = os.getenv('USER_AGENT')
if not user_agent:
    user_agent = that_default_value
```

tinez commented on api/services/real.py:
```diff
@@ -5,64 +5,98 @@
+                status=500,
+            )
+
+        return next(iter(places.values()))  # Get first available coordinates
```
Hmm why are we returning here only the first item from `places` and not the whole collection? Also, what is the type of the thing that we're returning here?

tinez commented on api/services/services.py:
```diff
@@ -12,6 +12,24 @@ def handle_address(street, house_number, radius):
+
+    # Check if Nominative service returned an error (JsonResponse)
+    # Use type name check to avoid import issues
+    if hasattr(address_coordinate, '__class__') and 'JsonResponse' in str(type(address_coordinate)):
```
Ok, so the API can return a `JsonResponse`, which means we're dealing with an error, or smth else. It would be good to reflect this in the type signature of `nominative_service.fetch_data`, e.g.:
```
def fetch_data(...) -> Places | JsonResponse:
```
It's also hard to tell that `JsonResponse` actually means an error.
To check if it's actually an error you should use `isinstance(address_coordinate, JsonResponse)`.
It's also a good idea to have a separate type to represent the "good" result (ideally you'd use smth like `pydantic` to validate the returned JSON and return a model object).

When designing APIs you usually go with one of these two approaches:
* raise an exception (preferrably one that you defined e.g. `DataRetrievalError` or smth) when a problem occurs and let the clients catch and handle it
* use a pattern of returning a proper value or an error like you did

Without any statistical data to prove it, I'd risk saying python projects tend to lean to the former ;)

tinez commented on api/views.py:
```diff
@@ -38,7 +57,18 @@ def analyze_address(request):
+        data_dir = Path(__file__).resolve().parent / 'data'
+        streets_path = data_dir / 'streets.json'
+        with streets_path.open(encoding='utf-8') as f:
+            street_data = json.load(f)
```
IIUC, this code loads the file on every request? You usually want to cache these kind of things. You could e.g. check the modification time on the file and read it only if it changed since last time.

tinez commented on deploy/docker-entrypoint.sh:
```diff
@@ -0,0 +1,34 @@
+export USE_MOCK_GISN="${USE_MOCK_GISN:-False}"
+
+echo "Starting TamaOd application..."
+echo "  SECRET_KEY: ${SECRET_KEY:0:10}..."
```
You should not log any parts of you secret key :)

tinez commented on scripts/runserver_local.sh:
```diff
@@ -0,0 +1,67 @@
+    if [ -f .env ]; then
+        ENV_FILE=.env DEBUG=True SECURE_SSL_REDIRECT=False SECURE_HSTS_ENABLE=False \
+            pdm run python manage.py runserver_plus \
+            --cert-file "$CERT_FILE" \
```
Can we use certs in production? There should be a way for the users to pass their own cert files into production deployment.

elai-shalev replied:
Will be handled in next PR
